File: cpp/graph_and_trees/invert_binary_tree.cpp
//
// Created by Harsha Vardhan on 27/06/22.
//

#include <iostream>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

void invertTreeHelper(TreeNode* root) {
    if (root == nullptr) return;
    TreeNode* tmp = root->left;
    root->left = root->right;
    root->right = tmp;
    invertTreeHelper(root->left);
    invertTreeHelper(root->right);
}

TreeNode* invertTree(TreeNode* root) {
    invertTreeHelper(root);
    return root;
}

int main() {
    return 0;
}

File: go/binary_tree/invert_binary_tree.go
package main

func invertTree(root *TreeNode) *TreeNode {
	invertTreeHelper(root)
	return root
}

func invertTreeHelper(root *TreeNode) {
	if root == nil {
		return
	}
	if root.Left == nil && root.Right == nil {
		return
	}
	root.Left, root.Right = root.Right, root.Left
	invertTreeHelper(root.Left)
	invertTreeHelper(root.Right)
}
